packages/instantsearch.js/src/connectors/related-products/connectRelatedProducts.ts

```diff
@@ -0,0 +1,150 @@
+  threshold?: number;
+  /**
+   * List of search parameters to send.
+   */
+  fallbackParameters?: Partial<
+    Omit<SearchParameters, 'page' | 'hitsPerPage' | 'offset' | 'length'>
+  >;
+  /**
+   * List of search parameters to send.
+   */
+  queryParameters?: Partial<
+    Omit<SearchParameters, 'page' | 'hitsPerPage' | 'offset' | 'length'>
+  >;
+  /**
+   * Function to transform the items passed to the templates.
+   */
+  transformItems?: TransformItems<Hit<THit>, { results: RecommendResultItem }>;
+};
+
+export type RelatedProductsWidgetDescription<THit extends BaseHit = BaseHit> = {
+  $$type: 'ais.relatedProducts';
+  renderState: RelatedProductsRenderState<THit>;
+};
+
+export type RelatedProductsConnector<THit extends BaseHit = BaseHit> =
+  Connector<
+    RelatedProductsWidgetDescription<THit>,
+    RelatedProductsConnectorParams<THit>
+  >;
+
+const connectRelatedProducts: RelatedProductsConnector =
+  function connectRelatedProducts(renderFn, unmountFn = noop) {
+    checkRendering(renderFn, withUsage());
+
+    return function relatedProducts(widgetParams) {
+      const {
+        objectID,
+        maxRecommendations,
+        threshold,
+        fallbackParameters,
+        queryParameters,
+        transformItems = ((items) => items) as NonNullable<
+          RelatedProductsConnectorParams['transformItems']
+        >,
+      } = widgetParams || {};
+
+      if (!objectID) {
+        throw new Error(withUsage('The `objectID` option is required.'));
+      }
+
+      return {
+        dependsOn: 'recommend',
+        $$type: 'ais.relatedProducts',
+
+        init(initOptions) {
+          renderFn(
+            {
+              ...this.getWidgetRenderState(initOptions),
+              instantSearchInstance: initOptions.instantSearchInstance,
+            },
+            true
+          );
+        },
+
+        render(renderOptions) {
+          const renderState = this.getWidgetRenderState(renderOptions);
+
+          renderFn(
+            {
+              ...renderState,
+              instantSearchInstance: renderOptions.instantSearchInstance,
+            },
+            false
+          );
+        },
+
+        getRenderState(renderState) {
+          return renderState;
+        },
+
+        getWidgetRenderState({ results }) {
+          if (results === null || results === undefined) {
+            return { recommendations: [], widgetParams };
+          }
+
+          return {
+            recommendations: transformItems(results.hits, {
+              results: results as RecommendResultItem,
+            }),
+            widgetParams,
+          };
+        },
+
+        dispose({ state }) {
+          unmountFn();
+
+          return state;
+        },
+
+        getWidgetParameters(state) {
+          return state.addRelatedProducts({
+            objectID,
+            maxRecommendations,
+            threshold,
+            fallbackParameters,
+            queryParameters,
+            $$id: this.$$id!,
```

Comment:
I'm not sure I see the point in first marking $$id as optional, but then in usage just non-null asserting it